metaflow/plugins/cards/card_datastore.py

```diff
@@ -153,25 +173,68 @@ def card_info_from_path(path):
 
     def save_card(self, card_type, card_html, card_id=None, overwrite=True):
         card_file_name = card_type
-        card_path = self.get_card_location(
-            self._get_card_path(), card_file_name, card_html, card_id=card_id
+        # HACK : FIXME (LATER) : Fix the duplication of below block in a few months.
+        # Check file blame to understand the age of this HACK.
+
+        # This function will end up saving cards at two locations.
+        # Thereby doubling the number of cards. (Which is a temporary fix)
+        # Why do this ? :
+        # When cards were introduced there was an assumption made about task-ids being unique.
+        # This assumption was incorrect.
+        # Only the pathspec needs to be unique but there is no such guarantees about task-ids.
+        # When task-ids are non-unique, card read would result in finding incorrect cards.
+        # This happens because cards were stored based on task-ids.
+        # If we immediately switch from storing based on task-ids to a step-name abstraction folder,
+        # then card reading will crash for many users.
+        # It would especially happen for users who are accessing cards created by a newer
+        # MF client from an older version of MF client.
+        # It will also easily end up breaking the metaflow-ui (which maybe using a client from an older version).
+        # Hence we are writing cards to both paths so that we can introduce breaking changes later in the future.
+        card_path_with_steps = self.get_card_location(
+            self._get_write_path(), card_file_name, card_html, card_id=card_id
         )
-        self._backend.save_bytes(
-            [(card_path, BytesIO(bytes(card_html, "utf-8")))], overwrite=overwrite
+        card_path_without_steps = self.get_card_location(
```

Comment:
As a small nit, this will compute the card hash twice. Not a major deal since it's a hack.

Reply:
Great catch! Since cards can dump a lot of information in a JSON, you can be hashing huge files. Will fix.